keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -0,0 +1,362 @@
+- Some CNI plugins are capable of assigning dual-stack addresses on a pod, but Kubernetes is aware of only one address per pod.
+- Service IPs are either all IPv4 or all IPv6 in a cluster.
+- Kubernetes system pods (api server, controller manager, etc.) can have only one IP address per pod, and addresses are either all IPv4 or all IPv6.
+- Endpoints for services are either all IPv4 or all IPv6.
+- Kube-dns is capable of running dual-stack, but it is currently only made aware of either all IPv4 or all IPv6 service addresses.
+
+For scenarios that require legacy IPv4-only clients or services, the above restrictions mean that complex and expensive IPv4/IPv6 transition mechanisms (e.g. NAT64/DNS64, stateless NAT46, or SIIT/MAP) will need to be implemented in the data center networking.
+
+An alternative is to allow for IPv4 and IPv6 pods and services to be supported in parallel throughout the cluster (a.k.a. full dual stack). This proposal aims to add full dual-stack access to Kubernetes clusters, providing native IPv4-to-IPv4 communication and native IPv6-to-IPv6 communication to, from and within a Kubernetes cluster.
+
+### Goals
+
+- Pod Connectivity: IPv4-to-IPv4 and IPv6-to-IPv6 access between pods
+- Service Access: IPv4-to-IPv4 and IPv6-to-IPv6 access from pods to Kubernetes services
+- External Server Access: IPv4-to-IPv4 and IPv6-to-IPv6 access from pods to external servers
+- Ingress Access: Access from IPv4 and/or IPv6 clients to Kubernetes services. Depending on the ingress controller used, connections may cross IP families (IPv4 clients to IPv6 services, or vice versa).
+- Service Discovery: Support for both A and AAAA records for kube-DNS and coreDNS
+- Functionality tested with the Bridge CNI plugin and Host-Local IPAM plugin as references
+- Maintain backwards-compatible support for IPv4-only and IPv6-only clusters
+
+### Non-Goals
+
+- Cross-family connectivity: IPv4-to-IPv6 and IPv6-to-IPv4 connectivity is considered outside of the scope of this proposal, except in the case of ingress access (depending on ingress controller used).
+- CNI network plugins other than the Bridge and Host-Local IPAM plugins should support Kubernetes dual stack, but the development and testing of dual stack support for these other plugins is considered outside of the scope of this proposal.
+- Code changes should support any number IPv4 addresses and any number of IPv6 addresses per pod (up to some reasonable limit). However, testing will be done only up to the following limits:
+  - Pod addresses: 1 IPv4 address and 2 IPv6 addresses per pod maximum
```

Comment:
^^ More reasons why it may be useful to call out use cases for dual-stack. One to identify the cases that could be addressed. Another is to see if the use cases **should** be addressed. And a third, to ensure we have an common understanding of how dual-stack is expected to be used.

Reply:
@thockin - I'll switch this to 1 IPv4 + 1 IPv6 for this KEP. The support for multiple IPv6 addresses seems to be generating a lot of questions, and I don't find the 1 ULA + 1 GUA that compelling of a user case. If someone comes up with a compelling use case, we can change the test limits.